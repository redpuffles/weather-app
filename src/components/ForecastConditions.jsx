/**
 * A component that contains and handles ForecastDay components for the next
 * five days (including the current day).
 * 
 * Processes relevant weatherData data from props and passes them down to each
 * children day component.
 * 
 * @file This file defines the ForecastConditions component class.
 * @author Roger.
 * @since 1.0.3
 */

import PropTypes from "prop-types";
import React from "react";

import ForecastDay from "./ForecastDay";

import * as S from "../styles/ForecastConditions.styles";

function ForecastConditions (props) {
  const [forecastData, setForecastData] = React.useState([]); // An array to hold each forecast day's data. Should be length five, ordered by day.

  /**
   * Processes props.weatherData and holds the relevant parts for each
   * ForecastDay component.
   * 
   * Gets relevant data for the first five days. If an error occurs, a message
   * is displayed on the home page.
   */
  React.useEffect(() => {
    var forecastDataNew = [];
    var i = 0;

    try {
      for (i = 0; i < 5; i++) {
        const date = new Date(props.weatherData.daily[i].dt * 1000);

        const data = {
          alt: props.weatherData.daily[i].weather[0].description,
          day: props.dayFromNum(date.getDay()),
          icon: props.weatherData.daily[i].weather[0].icon,
          key: i,
          temp: String(props.weatherData.daily[i].temp.day)
        }

        forecastDataNew = [ ...forecastDataNew, data ];
      }
    } catch (e) {
      props.handleHomeError("Couldn't fetch daily forecast data from the servers.");

      var date = new Date();

      for (i = 0; i < 5; i++) {
        const data = {
          alt: "",
          day: props.dayFromNum(date.getDay()),
          icon: "",
          key: i,
          temp: ""
        }

        forecastDataNew = [ ...forecastDataNew, data ];

        date.setDate(date.getDate() + 1);
      }
    } finally {
      setForecastData(forecastDataNew);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.weatherData])

  return (
    <S.ForecastConditions>
      {forecastData.map((i) =>
        <ForecastDay
          alt={i.alt}
          day={i.day}
          getIconURL={props.getIconURL}
          icon={i.icon}
          key={i.key}
          temp={i.temp}
        />
      )}
    </S.ForecastConditions>
  );
}

ForecastConditions.propTypes = {
  dayFromNum: PropTypes.func.isRequired,
  getIconURL: PropTypes.func.isRequired,
  handleHomeError: PropTypes.func.isRequired,
  weatherData: PropTypes.object
}

export default ForecastConditions;
