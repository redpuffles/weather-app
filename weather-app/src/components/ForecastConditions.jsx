import React from "react";

import ForecastDay from "./ForecastDay";

import * as S from "../styles/ForecastConditions.styles";

function ForecastConditions (props) {
  const [forecast, setForecast] = React.useState([]);

  React.useEffect(() => {
    const dayOfWeek = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
    var forecastTemp = [];

    const now = new Date();
    const nowDate = String(now.getDate()).padStart(2, "0");
    const nowMonth = new Intl.DateTimeFormat("en-US", { month: "short" }).format(now);
    const nowDay = dayOfWeek[now.getDay()];
    
    var dateDay;
    var i;

    try {
      for (i = 0; i < 5; i++) {
        const date = new Date(props.data.daily[i].dt * 1000);
        const dateDate = String(date.getDate()).padStart(2, "0");
        const dateMonth = new Intl.DateTimeFormat("en-US", { month: "short" }).format(date);
        dateDay = dayOfWeek[date.getDay()];

        if ((nowDate === dateDate) && (nowMonth === dateMonth) && (nowDay === dateDay)) {
          dateDay = "Today";
        }

        const data = {
          day: dateDay,
          description: props.data.daily[i].weather[0].description,
          icon: props.data.daily[i].weather[0].icon,
          temp: props.data.daily[i].temp.day
        }

        forecastTemp = [ ...forecastTemp, data ];
      }

      setForecast(forecastTemp);
    } catch (e) {
      var date = now;
      dateDay = nowDay;

      for (i = 0; i < 5; i++) {
        const data = {
          day: dateDay,
          description: "",
          icon: "",
          temp: ""
        }

        forecastTemp = [ ...forecastTemp, data ];

        date.setDate(date.getDate() + 1);
        dateDay = dayOfWeek[date.getDay()];
      }

      setForecast(forecastTemp);
    }
  }, [props]);

  return (
    <S.ForecastConditions>
      {forecast.map((item) =>
        <ForecastDay day={item.day} description={item.description} icon={item.icon} key={item.day} temp={item.temp} />
      )}
    </S.ForecastConditions>
  );
}

export default ForecastConditions;
