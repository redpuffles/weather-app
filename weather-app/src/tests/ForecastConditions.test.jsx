/**
 * React unit tests for ForecastConditions.
 * 
 * - "renders properly with the minimal required props"
 * - "calls handleHomeError if weatherData isn't provided"
 * 
 * @file This file contains unit tests for the ForecastConditions component.
 * @author Roger.
 * @since 1.0.0
 */

import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import renderer from 'react-test-renderer';
import { screen } from '@testing-library/react';

import ForecastConditions from "../components/ForecastConditions";

function dayFromNum (num) {
  const dayOfWeek = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
  return dayOfWeek[num];
}

let container = null;

afterEach(() => {
  // Clean up after exit.
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

beforeEach(() => {
  // Set up a DOM element as a render target.
  container = document.createElement("div");
  document.body.appendChild(container);
});

it("renders properly with the minimal required props", () => {
  // default values when weatherData isn't given as a prop. checks that five
  // children components are rendered, each with the correct day strings.
  act(() => {
    render(<ForecastConditions
              dayFromNum={dayFromNum}
              getIconURL={() => {}}
              handleHomeError={() => {}}
           />, container);
  });
  var date = new Date();
  for (var i = 0; i < 5; i++) {
    expect(screen.getByRole("forecast-day",
                            { name: dayFromNum(date.getDay()) })).toBeTruthy();
    date.setDate(date.getDate() + 1);
  }

  // snapshot rendering with minimal props
  const test = renderer.create(<ForecastConditions
                                  dayFromNum={(n) => { return "Mon"; }}
                                  getIconURL={() => {}}
                                  handleHomeError={() => {}}
                               />, container);
  expect(test).toMatchSnapshot();
});

it("calls handleHomeError if weatherData isn't provided", () => {
  const mockFunc = jest.fn();

  act(() => {
    render(<ForecastConditions
              dayFromNum={(n) => { return "Mon"; }}
              getIconURL={() => {}}
              handleHomeError={mockFunc}
           />, container);
  });
  expect(mockFunc).toHaveBeenCalledTimes(1);
});
