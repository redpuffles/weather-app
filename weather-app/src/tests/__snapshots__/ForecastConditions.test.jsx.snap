// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly with the minimal required props 1`] = `
<div
  className="sc-eCApnc kEfoWn"
/>
`;
